### PR TITLE
drivers: watchdog: TI CC13xx/CC26xx watchdog timer support

### LINKED ISSUE #39331: ti: cc13xx_cc26xx: watchdog timer driver
**Is your enhancement proposal related to a problem? Please describe.**
A watchdog driver for TI CC13xx / CC26xx MCUs would be nice :D  

**Describe the solution you'd like**
Implementation in upcoming PR...

### PR DESCRIPTION
**Pull request by niflostancu**
Implemented new WDT driver for TI CC13xx/CC26xx family using HAL TI routines.
Supports interrupts & MCU soft reset on timeout.

Follow up from stale / closed PR #39332 :(
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/39331

Tested on my CC1352P1 LaunchXL using samples/drivers/watchdog.

### CODE REVIEW

File: drivers/watchdog/wdt_cc13xx_cc26xx.c [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+/* Driverlib includes */
+#include <inc/hw_types.h>
+#include <inc/hw_wdt.h>
+#include <driverlib/rom.h>
+#include <driverlib/prcm.h>
+#include <driverlib/watchdog.h>
+
+#define CPU_FREQ DT_PROP(DT_PATH(cpus, cpu_0), clock_frequency)
+
+#define WDT_INST_IDX            0
```
Comment (mbolivar-nordic) [inline drivers/watchdog/wdt_cc13xx_cc26xx.c:26]: ?

File: dts/bindings/watchdog/ti,cc13xx-cc26xx-watchdog.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+compatible: "ti,cc13xx-cc26xx-watchdog"
+
+include: base.yaml
+
+properties:
+    reg:
+      required: true
+
+    label:
+      required: true
```
Comment (mbolivar-nordic) [inline dts/bindings/watchdog/ti,cc13xx-cc26xx-watchdog.yaml:15]: Remove label property. We are moving away from them.

File: dts/arm/ti/cc13x2_cc26x2.dtsi [hunk 3]
```diff
@@ -149,6 +155,14 @@
 				status = "disabled";
 			};
 		};
+
+		wdt0: watchdog@40080000 {
+			compatible = "ti,cc13xx-cc26xx-watchdog";
+			reg = <0x40080000 0x1000>;
+			interrupts = <EXP_WDT 0>;
+			label = "WDT_0";
```
Comment (mbolivar-nordic) [inline dts/arm/ti/cc13x2_cc26x2.dtsi:163]: Remove

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

File: drivers/watchdog/wdt_cc13xx_cc26xx.c (hunk 1 above)
Comment (niflostancu) [reply drivers/watchdog/wdt_cc13xx_cc26xx.c:26]: I needed the system clock to convert milliseconds to WDT reload ticks.
I don't remember where I got this, most probably from here: [`log_backend_swo.c`](https://github.com/zephyrproject-rtos/zephyr/blob/f249f278bf586c87b89b7360d65977d71a236f21/subsys/logging/log_backend_swo.c#L49).
Is there another / preferred way to read the system clock?

As for `WDT_INST_IDX`, I copied some driver initialization code which used `DT_INST_FOREACH_*` and I renamed the index parameter to this.
Should I expand the constant `0` everywhere or should I rewrite it using `DT_INST_FOREACH_*` regardless of the WDT being a singleton? (defined code is pretty hard to read :D)

Thanks.

File: dts/bindings/watchdog/ti,cc13xx-cc26xx-watchdog.yaml (hunk 2 above)
Comment (niflostancu) [reply dts/bindings/watchdog/ti,cc13xx-cc26xx-watchdog.yaml:15]: yes, I saw this inside git history but totally forgot to remove them.

File: drivers/watchdog/wdt_cc13xx_cc26xx.c (hunk 1 above)
Comment (cfriedt) [reply drivers/watchdog/wdt_cc13xx_cc26xx.c:26]: Yes, please use `DT_INST_FOREACH_*()`

Comment (niflostancu) [reply drivers/watchdog/wdt_cc13xx_cc26xx.c:26]: Okay, I'll get to it tomorrow ;)

**Comment by pepe2k** [comment]
@cfriedt @niflostancu don't want to hijack this PR but I was working on the same driver few weeks ago, there are still some TODOs (see the commit message for the driver code) but maybe you could make use of it or part of it? I pushed my version here: https://github.com/pepe2k/zephyr/tree/ti-cc13xx-cc26xx_watchdog-driver

Cheers,
Piotr

**Comment by niflostancu** [comment]
Hi, thank you for your suggestions.
Based on them, I've done some extensive modifications to the PR:

- rewritten init code to use `DT_INST_FOREACH_STATUS_OKAY`;
- implemented NMI mode (using a new boolean DTSI property: `interrupt-nmi`; hope that's the best way to do it);
- replaced msToTicks conversion function with a macro;
- move the `_enable()` code inside `_setup()`;
- if WDT is enabled at boot, call `_setup()` on init with the default option `WDT_OPT_PAUSE_HALTED_BY_DBG` based on `CONFIG_DEBUG`;
- enabled WDT feature for all supported boards (in separate commit, incl. doc updates);
- better comments explaining the implementation decisions made.

Note: as various TI forum posts suggest [1] (I couldn't find any explicit mentions in the Reference), the watchdog clock is always paused during sleep, so there is  nothing to be done in this case (I avoided any option checks for this to be permissive).

Since I based my solution on some of @pepe2k's code, how shall we proceed with the copyright notice in the header? Shall I insert his name + email too, separated by a comma?

Thanks.

[1] https://e2e.ti.com/support/wireless-connectivity/sub-1-ghz-group/sub-1-ghz/f/sub-1-ghz-forum/990138/cc1312r-hw-watchdog-strange-behavior-in-debug/3666163#3666163

**Comment by pepe2k** [comment]
Hi @niflostancu

> rewritten init code to use `DT_INST_FOREACH_STATUS_OKAY`;

I see that this was suggested here by @cfriedt and you did it also in your another PR #49229 but I really don't understand this approach. From what I understand, this in result allows to generate several instances of the same driver for separated blocks of the same hardware, based on the DTS. But... you are still using the TI's driverlib (at least for the wdt that's only incline functions accessing some registers, not sure about the ADC if there is anything already in the ROM) which doesn't even allow to pass register base to functions. Plus, in case of this series of SOCs, there is just one watchdog and I doubt there will be anything new with more than one. So in this case, we have some extra code, e.g. the `reg` which is not used at all.
@cfriedt could you please explain motivation behind this suggestion to use `DT_INST_FOREACH_STATUS_OKAY` instead of `DEVICE_DT_INST_DEFINE`?

> * implemented NMI mode (using a new boolean DTSI property: `interrupt-nmi`; hope that's the best way to do it);

I was planning to handle it as a Kconfig option. No opinion on the DTS property but I'm pretty sure that introducing something like this would need to be first discussed with a wider audience and defined somewhere in bindings for the `watchdog` subsystem.

> Note: as various TI forum posts suggest [1] (I couldn't find any explicit mentions in the Reference), the watchdog clock is always paused during sleep, so there is nothing to be done in this case (I avoided any option checks for this to be permissive).

I found it in datasheet (not the reference manual):

> The watchdog timer is used to regain control if the system operates incorrectly due to software errors. It is typically used to generate an interrupt to and reset of the device for the case where periodic monitoring of the system components and tasks fails to verify proper functionality. The watchdog timer runs on a 1.5 MHz clock rate and cannot be stopped once enabled. **The watchdog timer pauses to run in Standby power mode and when a debugger halts the device**.

Watchdog is also listed under `Power Management` in `Power Modes` table in this document. That's why I have some doubts and still plan to do more tests.
Source: https://www.ti.com/lit/gpn/cc1352r

I have also some additional comments, will add them in code.

Cheers,
Piotr

File: drivers/watchdog/Kconfig.cc13xx_cc26xx [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) 2021 Florin Stancu <niflostancu@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_TI_CC13XX_CC26XX_WDG := ti,cc13xx-cc26xx-watchdog
+
+config WDT_CC13XX_CC26XX
+	bool "Watchdog Driver for CC13xx / CC26xx family of MCUs"
+	default $(dt_compat_enabled,$(DT_COMPAT_TI_CC13XX_CC26XX_WDG))
```
Comment (pepe2k) [inline drivers/watchdog/Kconfig.cc13xx_cc26xx:8]: How about `DT_HAS_<COMPAT>_ENABLED` and drop both below `depends`, see e.g. bdff3e76e5853fdc9dce2e72858dd9936b90d8fa

File: drivers/watchdog/wdt_cc13xx_cc26xx.c [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+#include <zephyr/drivers/watchdog.h>
+#include <soc.h>
+#include <errno.h>
+
+#define LOG_LEVEL CONFIG_WDT_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(wdt_cc13xx_cc26xx);
+
+/* Driverlib includes */
```
Comment (pepe2k) [inline drivers/watchdog/wdt_cc13xx_cc26xx.c:17]: Isn't just `<driverlib/watchdog.h>` enough?

File: drivers/watchdog/wdt_cc13xx_cc26xx.c [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+	return 0;
+}
+
+static int wdt_cc13xx_cc26xx_setup(const struct device *dev, uint8_t options)
+{
+	const struct wdt_cc13xx_cc26xx_cfg *config = dev->config;
+	struct wdt_cc13xx_cc26xx_data *data = dev->data;
+
+	/*
+	 * Note: don't check if watchdog is already enabled, an application might
```
Comment (pepe2k) [inline drivers/watchdog/wdt_cc13xx_cc26xx.c:87]: I have some doubts here, see: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/watchdog.h#L161

We allow to `disable` this watchdog (but only by disabling generating the reset) so the `setup` can be called again. But is this a correct approach? In fact, this type of watchdog cannot be stopped.

Comment (niflostancu) [reply drivers/watchdog/wdt_cc13xx_cc26xx.c:83]: Yeah, the primary use case for this was to be able change the timeout dynamically.
A Zephyr Watchdog API behavior change like being able to call `install_timeout` after setup would be a better way, though...

Comment (niflostancu) [reply drivers/watchdog/wdt_cc13xx_cc26xx.c:83]: Or, we could further _emulate_ the watchdog being stopped by defining a boolean flag that gets set when `wdt_disable` is called then preventing the execution of the user callback (note: the interrupt handler will still fire).

**Review by cfriedt — DISMISSED** [review]

**Comment by martinjaeger** [comment]
@niflostancu can you please rebase to fix the merge conflicts?

**Comment by cfriedt** [comment]
> @cfriedt could you please explain motivation behind this suggestion to use `DT_INST_FOREACH_STATUS_OKAY` instead of `DEVICE_DT_INST_DEFINE`?
 
`DEVICE_DT_INST_DEFINE` does not take status into account, while `DT_INST_FOREACH_STATUS_OKAY` does.

Alternatively, `DEVICE_DT_GET_ANY` or `DEVICE_DT_GET_ONE`.

**Review by martinjaeger — DISMISSED** [review]

**Review by mbolivar-nordic — DISMISSED** [review]

**Comment by cfriedt** [comment]
Why is this failing unrelated tests? Has it been rebased properly?

**Comment by niflostancu** [comment]
It's not me :( 

It fails on `main`, too:
```
>> git rev-parse HEAD
e8f0e66bc4281fec4f7489b8fc54f488761ab91c

>> west build --pristine -b s32z270dc2_rtu1_r52 samples/drivers/watchdog/
[...]
[1/180] Preparing syscall dependency handling
[2/180] Generating include/generated/version.h
-- Zephyr version: 3.2.99 (zephyr), build: zephyr-v3.2.0-3143-ge8f0e66bc428
[168/180] Building C object CMakeFiles/app.dir/src/main.c.obj
FAILED: CMakeFiles/app.dir/src/main.c.obj
ccache zephyr-sdk/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DKERNEL -D__PROGRAM_START -D__ZEPHYR__=1 -Izephyr/include -Izephyr/build/zephyr/include/generated -Izephyr/soc/arm/nxp_s32/s32ze -Izephyr/soc/arm/nxp_s32/common/. -Imodules/hal/cmsis/CMSIS/Core_R/Include -Imodules/hal/nxp/s32/drivers/s32ze/BaseNXP/header -Imodules/hal/nxp/s32/drivers/s32ze/BaseNXP/include -Imodules/hal/nxp/s32/drivers/s32ze/Mcu/include -Imodules/hal/nxp/s32/drivers/s32ze/Port/include -Imodules/hal/nxp/s32/drivers/s32ze/Rte/include -Imodules/hal/nxp/s32/drivers/s32ze/Uart/include -Imodules/hal/nxp/s32/drivers/s32ze/Swt/include -Imodules/hal/nxp/s32/soc/s32z27/include -Izephyr/modules/hal_nxp/. -isystem zephyr/lib/libc/minimal/include -isystem zephyr-sdk-0.15.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/include -isystem zephyr-sdk-0.15.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/include-fixed -fno-strict-aliasing -Os -imacros zephyr/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-r52 -mabi=aapcs -mfpu=neon-fp-armv8 -mfloat-abi=hard -mfp16-format=ieee --sysroot=zephyr-sdk/arm-zephyr-eabi/arm-zephyr-eabi -imacros zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=zephyr/samples/drivers/watchdog=CMAKE_SOURCE_DIR -fmacro-prefix-map=zephyr=ZEPHYR_BASE -fmacro-prefix-map=zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT CMakeFiles/app.dir/src/main.c.obj -MF CMakeFiles/app.dir/src/main.c.obj.d -o CMakeFiles/app.dir/src/main.c.obj -c zephyr/samples/drivers/watchdog/src/main.c
In file included from zephyr/include/zephyr/toolchain/gcc.h:89,
                 from zephyr/include/zephyr/toolchain.h:50,
                 from zephyr/include/zephyr/kernel_includes.h:19,
                 from zephyr/include/zephyr/kernel.h:17,
                 from zephyr/samples/drivers/watchdog/src/main.c:9:
zephyr/samples/drivers/watchdog/src/main.c: In function 'main':
zephyr/include/zephyr/device.h:83:41: error: '__device_dts_ord_DT_N_ALIAS_watchdog0_ORD' undeclared (first use in this function)
   83 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                         ^~~~~~~~~
[...]
zephyr/samples/drivers/watchdog/src/main.c:73:42: note: in expansion of macro 'DEVICE_DT_GET'
   73 |         const struct device *const wdt = DEVICE_DT_GET(DT_ALIAS(watchdog0));
      |                                          ^~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build zephyr/build
```

**Comment by cfriedt** [comment]
> It's not me :(
> 
> It fails on `main`, too:
> 
> ```
> >> git rev-parse HEAD
> e8f0e66bc4281fec4f7489b8fc54f488761ab91c
> 
> >> west build --pristine -b s32z270dc2_rtu1_r52 samples/drivers/watchdog/
> [...]
> [1/180] Preparing syscall dependency handling
> [2/180] Generating include/generated/version.h
> -- Zephyr version: 3.2.99 (zephyr), build: zephyr-v3.2.0-3143-ge8f0e66bc428
> [168/180] Building C object CMakeFiles/app.dir/src/main.c.obj
> FAILED: CMakeFiles/app.dir/src/main.c.obj
> ccache zephyr-sdk/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DKERNEL -D__PROGRAM_START -D__ZEPHYR__=1 -Izephyr/include -Izephyr/build/zephyr/include/generated -Izephyr/soc/arm/nxp_s32/s32ze -Izephyr/soc/arm/nxp_s32/common/. -Imodules/hal/cmsis/CMSIS/Core_R/Include -Imodules/hal/nxp/s32/drivers/s32ze/BaseNXP/header -Imodules/hal/nxp/s32/drivers/s32ze/BaseNXP/include -Imodules/hal/nxp/s32/drivers/s32ze/Mcu/include -Imodules/hal/nxp/s32/drivers/s32ze/Port/include -Imodules/hal/nxp/s32/drivers/s32ze/Rte/include -Imodules/hal/nxp/s32/drivers/s32ze/Uart/include -Imodules/hal/nxp/s32/drivers/s32ze/Swt/include -Imodules/hal/nxp/s32/soc/s32z27/include -Izephyr/modules/hal_nxp/. -isystem zephyr/lib/libc/minimal/include -isystem zephyr-sdk-0.15.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/include -isystem zephyr-sdk-0.15.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/include-fixed -fno-strict-aliasing -Os -imacros zephyr/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-r52 -mabi=aapcs -mfpu=neon-fp-armv8 -mfloat-abi=hard -mfp16-format=ieee --sysroot=zephyr-sdk/arm-zephyr-eabi/arm-zephyr-eabi -imacros zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=zephyr/samples/drivers/watchdog=CMAKE_SOURCE_DIR -fmacro-prefix-map=zephyr=ZEPHYR_BASE -fmacro-prefix-map=zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT CMakeFiles/app.dir/src/main.c.obj -MF CMakeFiles/app.dir/src/main.c.obj.d -o CMakeFiles/app.dir/src/main.c.obj -c zephyr/samples/drivers/watchdog/src/main.c
> In file included from zephyr/include/zephyr/toolchain/gcc.h:89,
>                  from zephyr/include/zephyr/toolchain.h:50,
>                  from zephyr/include/zephyr/kernel_includes.h:19,
>                  from zephyr/include/zephyr/kernel.h:17,
>                  from zephyr/samples/drivers/watchdog/src/main.c:9:
> zephyr/samples/drivers/watchdog/src/main.c: In function 'main':
> zephyr/include/zephyr/device.h:83:41: error: '__device_dts_ord_DT_N_ALIAS_watchdog0_ORD' undeclared (first use in this function)
>    83 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
>       |                                         ^~~~~~~~~
> [...]
> zephyr/samples/drivers/watchdog/src/main.c:73:42: note: in expansion of macro 'DEVICE_DT_GET'
>    73 |         const struct device *const wdt = DEVICE_DT_GET(DT_ALIAS(watchdog0));
>       |                                          ^~~~~~~~~~~~~
> ninja: build stopped: subcommand failed.
> FATAL ERROR: command exited with status 1: /usr/bin/cmake --build zephyr/build
> ```

I feel like I've seen errors like this before when a DT node does not have `status = "okay";` @niflostancu - can you double check?

**Comment by niflostancu** [comment]
Please see #54000 for the actual issue. Apparently, the CI skipped this test when `s32z270dc2` watchdog support was added (the broken commit is probably e63d1389a9 ).

**Review by cfriedt — DISMISSED** [review]

**Comment by martinjaeger** [comment]
@niflostancu this needs another rebase to resolve the merge conflicts.

**Review by martinjaeger — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]

**Review by martinjaeger — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Comment by cfriedt** [comment]
Just in case anyone comes asking, this PR was merged because it has had no code changes in months but has been rebased over and over to e.g. avoid unrelated CI failures, merge conflicts, and so on.

It's been open for a year (if not this specific PR then a stale / closed prior version). I don't want to force @niflostancu to defer any longer.

Feel free to publicly shame me / force me to fix all of the bugs if this breaks the rules.

**Comment by stephanosio** [comment]
While I understand the circumstances involved and the frustration it may have caused, the feature merge freeze process is there for a reason. If we keep making exceptions on an arbitrary basis, it will become harder to draw a line on what should go in after the feature freeze, and will eventually result in the loss of any benefits of implementing the feature freeze/release candidate steps in the release process.

IMHO, this should have been evaluated and approved by the TSC/CCB as per the documented process.

**Comment by niflostancu** [comment]
Hi, thanks all for approving this.

I understand your point and especially the fact that large projects like this need very firm procedures and checks to avoid bugs / bloat / other issues.
But I also agree with @cfriedt (whom I thank for the support) and find that needlessly stretching this process for very long periods is a a bit close to the other extreme. 
<_also insert your favorite "rules are to be broken" quote ;)_ >

I realize that the `TI CCxx` platform has few stakeholders and probably no maintainers and this might incur some delays, though this PR introduces an isolated module (watchdog driver for a very specific platform) that cannot really affect (e.g., introduce bugs) anywhere else.

Regarding my contributions, I assure you they are thoroughly tested on real hardware (I already use them in production for over a year) and I am 100% available for fixing any bugs discovered.

**Pull request merged**